scripts/build/setup/termux_setup_golang.sh

```diff
@@ -19,20 +19,28 @@ termux_setup_golang() {
 		termux_error_exit "Unsupported arch: $TERMUX_ARCH"
 	fi
 
-	local TERMUX_GO_VERSION=go1.12.7
-	local TERMUX_GO_PLATFORM=linux-amd64
+	if [ -z "$TERMUX_ON_DEVICE_BUILD" ]; then
+		local TERMUX_GO_VERSION=go1.12.7
+		local TERMUX_GO_PLATFORM=linux-amd64
 
-	local TERMUX_BUILDGO_FOLDER=$TERMUX_COMMON_CACHEDIR/${TERMUX_GO_VERSION}
-	export GOROOT=$TERMUX_BUILDGO_FOLDER
-	export PATH=$GOROOT/bin:$PATH
+		local TERMUX_BUILDGO_FOLDER=$TERMUX_COMMON_CACHEDIR/${TERMUX_GO_VERSION}
+		export GOROOT=$TERMUX_BUILDGO_FOLDER
+		export PATH=$GOROOT/bin:$PATH
 
-	if [ -d "$TERMUX_BUILDGO_FOLDER" ]; then return; fi
+		if [ -d "$TERMUX_BUILDGO_FOLDER" ]; then return; fi
 
-	local TERMUX_BUILDGO_TAR=$TERMUX_COMMON_CACHEDIR/${TERMUX_GO_VERSION}.${TERMUX_GO_PLATFORM}.tar.gz
-	rm -Rf "$TERMUX_COMMON_CACHEDIR/go" "$TERMUX_BUILDGO_FOLDER"
-	termux_download https://storage.googleapis.com/golang/${TERMUX_GO_VERSION}.${TERMUX_GO_PLATFORM}.tar.gz \
-		"$TERMUX_BUILDGO_TAR" \
-		66d83bfb5a9ede000e33c6579a91a29e6b101829ad41fffb5c5bb6c900e109d9
+		local TERMUX_BUILDGO_TAR=$TERMUX_COMMON_CACHEDIR/${TERMUX_GO_VERSION}.${TERMUX_GO_PLATFORM}.tar.gz
+		rm -Rf "$TERMUX_COMMON_CACHEDIR/go" "$TERMUX_BUILDGO_FOLDER"
+		termux_download https://storage.googleapis.com/golang/${TERMUX_GO_VERSION}.${TERMUX_GO_PLATFORM}.tar.gz \
+			"$TERMUX_BUILDGO_TAR" \
+			66d83bfb5a9ede000e33c6579a91a29e6b101829ad41fffb5c5bb6c900e109d9
 
-	( cd "$TERMUX_COMMON_CACHEDIR"; tar xf "$TERMUX_BUILDGO_TAR"; mv go "$TERMUX_BUILDGO_FOLDER"; rm "$TERMUX_BUILDGO_TAR" )
+		( cd "$TERMUX_COMMON_CACHEDIR"; tar xf "$TERMUX_BUILDGO_TAR"; mv go "$TERMUX_BUILDGO_FOLDER"; rm "$TERMUX_BUILDGO_TAR" )
+	else
+		if [ "$(dpkg-query -W -f '${db:Status-Status}\n' golang 2>/dev/null)" != "installed" ]; then
+			# For on-device builds don't care about which version of golang is installed.
+			# In most cases it should be suitable.
+			apt update && apt install -y golang
```

Comment:
Can we make all these package installs optional? So the user will be prompted and can stop the build process if some package is unwanted.

Reply:
There is no point in this check. If user want to build golang package, then it should be installed.

Another variant will be to download and unpack deb file to install golang into somewhere in `$TERMUX_TOPDIR/_cache/` like done when cross-compiling.

Reply:
Wouldn't it be as simple as removing the -y here or is more needed to stop the build? I think it should be transparent to the dev what packages are being installed, as we will now be installing packages only to build other packages. For example, I will be disabling the build-essential package install above, as I don't want a ton of build-related packages that I don't use.

Reply:
Just removing `-y` option is not enough. If package dependencies are already installed, then `apt install {package name}` will not prompt for yes/no.

Reply:
> For example, I will be disabling the build-essential package install above, as I don't want a ton of build-related packages that I don't use.

Each of `termux_setup_cmake.sh` `termux_setup_golang.sh` `termux_setup_meson.sh` `termux_setup_ninja.sh` `termux_setup_protobuf.sh` `termux_setup_rust.sh` scripts are used only when building package that requires such utilities.

If you will build only packages with which you are working, you probably never get those tools installed unless you source these scripts inside build.sh of your package.

Reply:
OK, removing -y as you just did is probably enough for all those, but I'd appreciate if you made installing build-essential optional, for those of us who know what we're doing and only want a subset of those packages installed.

Reply:
> removing -y as you just did is probably enough for all those

Is not enough. Golang is getting installed without prompt.

> but I'd appreciate if you made installing build-essential optional

It contain **minimal** dependencies for ./build-package.sh. It won't be optional.

Reply:
> Is not enough. Golang is getting installed without prompt.

That's because its dependencies are already there though, right? What I was saying is that since it will only get installed if someone wants to build a package written in Go, as you pointed out, it's good enough if it simply prompts when the Go install needs more dependencies. If you mean the build script doesn't stop even if the user says no to installing Go, maybe that should be done also.

> It contain minimal dependencies for ./build-package.sh. It won't be optional.

I understand that many of those are practically required, but if I have no interest in building autoconf-using packages, I'd rather it not be installed. I'm simply asking to prompt the user before installing so many build packages, and let someone like me say no and move forward, as long as I'm willing to face the consequences that some dependency later on will not be met because I chose not to install the build-essential metapackage.

I never install such metapackages on Ubuntu and other distros, as I prefer to only install the build packages I need for my compilation.

Reply:
> If you mean the build script doesn't stop even if the user says no to installing Go, maybe that should be done also.

I mean that apt will not ask at all. And it really don't ask when only one (!) package is specified for installation.

> I understand that many of those are practically required, but if I have no interest in building autoconf-using packages, I'd rather it not be installed.

In such case then it will be just print warning that "package build-essential is not installed".

Reply:
> I mean that apt will not ask at all. And it really don't ask when only one (!) package is specified for installation.

I think it will ask as long as that one package requires other uninstalled packages, as you noted. What I don't know is if the bash script will try to keep going even if I say no to installing the Go package and its dependencies, which I'd rather it didn't.

> Will replace it with script setup-termux.sh, similar setup-ubuntu.sh. And it should be executed separately from build-package.sh.

Sounds good, thanks.

Reply:
What's done now:

* termux_setup_*.sh no longer uses `apt`. if package is not installed, it stops build and asks to install necessary packages.

* build-essential is no longer used. You will need to either install utilities manually. I.e. if got "NAME not found" error, then you probably need to install package containing utility with NAME. Or you can just install whole set with executing `scripts/setup-termux.sh`.

Reply:
Super, thanks for changing it. :smile: